src/time_utils.c

```diff
@@ -0,0 +1,51 @@
+#include "config.h"
+#include "time_utils.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/time.h>
+#include <time.h>
+
+void
+time_to_next_round_interval (const struct timeval now,
+                             const unsigned int interval,
+                             struct timeval *next)
+{
+  /* this is the current second in the current day */
+  time_t second_in_day = now.tv_sec - now.tv_sec / 86400 * 86400;
+
+  /* this determine's the next round interval by first dividing by the interval
+   * then multiplying which gives us the current interval start, then adding
+   * the interval which gives us the next interval start */
+  time_t next_interval = second_in_day / interval * interval + interval;
+
+  /* next we'll subtract to get seconds to next interval (also subtracting
+   * one as we'll also be checking millis below */
+  time_t seconds_to_next_interval = next_interval - second_in_day - 1;
+
+  /* do a similar calculation to determine approximate micros until next round
+   * interval */
+  unsigned long long now_millis = (((long long)now.tv_sec) * 1000LL)
+                                   + (long long)(now.tv_usec/1000LL);
+  unsigned long long micros_to_next_second =
+    (((now_millis / 1000LL) * 1000LL + 1000LL) - now_millis) * 1000LL;
```

Comment:
If 0≤now.tv_usec<1000, we get micros_to_next_second=1000000, which is out of range for the timer call.  We should either change the above calculation to avoid that or else correct it after the fact (incrementing next.tv_sec).